File: go/sync/internal/client/lease.go
package client

import (
	"context"
	"sync"
	"time"

	"github.com/fraym/freym-api/go/proto/sync/managementpb"
	"github.com/fraym/freym-api/go/sync/internal/util"
	"github.com/fraym/golog"
	"github.com/samber/lo"
)

type lock struct {
	tenant   string
	resource []string
}

type lease struct {
	ctx         context.Context
	stopWg      *sync.WaitGroup
	mx          *sync.RWMutex
	cancelLease context.CancelFunc
	leaseId     string
	locks       []lock
	rLocks      []lock

	appPrefix string
	ttl       int32
	ownIp     string
	client    managementpb.ServiceClient
	logger    golog.Logger
	onRenew   func()
}

func NewLease(
	ctx context.Context,
	appPrefix string,
	ownIp string,
	client managementpb.ServiceClient,
	logger golog.Logger,
) *lease {
	stopWg := &sync.WaitGroup{}
	stopWg.Add(1)

	return &lease{
		ctx:         ctx,
		stopWg:      stopWg,
		mx:          &sync.RWMutex{},
		cancelLease: nil,
		leaseId:     "",
		locks:       nil,
		rLocks:      nil,
		appPrefix:   appPrefix,
		ttl:         20,
		ownIp:       ownIp,
		client:      client,
		logger:      logger,
		onRenew:     nil,
	}
}

func (l *lease) OnRenew(fn func()) {
	l.mx.Lock()
	defer l.mx.Unlock()

	l.onRenew = fn
}

func (l *lease) LeaseId() string {
	l.mx.RLock()
	defer l.mx.RUnlock()

	return l.leaseId
}

func (l *lease) WaitForStop() {
	l.stopWg.Wait()
}

func (l *lease) getRegisteredLocksForPb() []*managementpb.Lock {
	if len(l.locks) == 0 {
		return nil
	}

	return lo.Map(l.locks, func(lock lock, _ int) *managementpb.Lock {
		return managementpb.Lock_builder{
			LeaseId:  l.leaseId,
			TenantId: lock.tenant,
			Resource: lock.resource,
		}.Build()
	})
}

func (l *lease) getRegisteredRLocksForPb() []*managementpb.Lock {
	if len(l.rLocks) == 0 {
		return nil
	}

	return lo.Map(l.rLocks, func(lock lock, _ int) *managementpb.Lock {
		return managementpb.Lock_builder{
			LeaseId:  l.leaseId,
			TenantId: lock.tenant,
			Resource: lock.resource,
		}.Build()
	})
}

func (l *lease) Renew() error {
	l.mx.Lock()
	defer l.mx.Unlock()

	ctx, cancel := context.WithCancel(context.Background())
	l.cancelLease = cancel

	leaseResponse, err := l.client.CreateLease(l.ctx, managementpb.CreateLeaseRequest_builder{
		Ip:                      l.ownIp,
		App:                     l.appPrefix,
		Ttl:                     l.ttl,
		LeaseId:                 l.leaseId,
		AlreadyRegisteredLocks:  l.getRegisteredLocksForPb(),
		AlreadyRegisteredRlocks: l.getRegisteredRLocksForPb(),
	}.Build())
	if err != nil {
		return err
	}

	l.leaseId = leaseResponse.GetLeaseId()

	go l.keepLeaseAlive(ctx)

	l.logger.Debug().Write("renewed lease")
	return nil
}

func (l *lease) keepLeaseAlive(leaseCtx context.Context) {
	ticker := time.NewTicker(time.Duration(l.ttl/2) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-leaseCtx.Done():
			l.mx.Lock()
			l.cancelLeaseIfExists()
			l.mx.Unlock()
			return
		case <-l.ctx.Done():
			_, err := l.client.DropLease(context.Background(), managementpb.DropLeaseRequest_builder{
				LeaseId: l.LeaseId(),
			}.Build())
			if err != nil {
				l.logger.Warn().WithError(err).Write("unable to drop lease")
			}
			l.stopWg.Done()
			l.mx.Lock()
			l.cancelLeaseIfExists()
			l.mx.Unlock()
			return
		case <-ticker.C:
			_, err := l.client.KeepLease(leaseCtx, managementpb.KeepLeaseRequest_builder{
				LeaseId: l.LeaseId(),
				Ttl:     l.ttl,
			}.Build())
			if err != nil {
				l.logger.Warn().WithError(err).Write("unable to keep lease")

				l.mx.Lock()
				l.onRenew()
				l.cancelLeaseIfExists()
				l.mx.Unlock()

				if err := util.Retry(l.Renew, time.Millisecond*100, 50); err != nil {
					l.logger.Fatal().WithError(err).Write("unable to renew lease")
				}
				return
			}
		}
	}
}

func (l *lease) cancelLeaseIfExists() {
	if l.cancelLease != nil {
		l.cancelLease()
		l.cancelLease = nil
	}
}

func (l *lease) Track(tenant string, resource []string, read bool) {
	l.mx.Lock()
	defer l.mx.Unlock()

	if read {
		l.rLocks = append(l.rLocks, lock{
			tenant:   tenant,
			resource: resource,
		})
		return
	}

	l.locks = append(l.locks, lock{
		tenant:   tenant,
		resource: resource,
	})
}

func (l *lease) Untrack(tenant string, resource []string, read bool) {
	l.mx.Lock()
	defer l.mx.Unlock()

	if read {
		_, index, found := lo.FindIndexOf(l.rLocks, func(rLock lock) bool {
			return rLock.tenant == tenant && util.SlicesEqual(rLock.resource, resource)
		})

		if found {
			l.rLocks[index] = l.rLocks[len(l.rLocks)-1]
			l.rLocks = l.rLocks[:len(l.rLocks)-1]
		}
		return
	}

	l.locks = lo.Filter(l.locks, func(lock lock, _ int) bool {
		return lock.tenant != tenant || !util.SlicesEqual(lock.resource, resource)
	})
}

File: js/sync/client/lease.ts
import { Lock as PbLock, ServiceClient } from "@fraym/proto/dist/index.freym.sync.management";
import { E_CANCELED, Mutex } from "async-mutex";
import { ServiceError } from "@grpc/grpc-js";
import { ClientConfig } from "./config";
import { Connection } from "./connection";
import { getOwnIpAddress } from "./ip";
import { createResolvablePromise, racePromises, wait } from "./promise";
import { retry } from "./retry";

export interface Lease {
    stop: () => void;
    runWithLeaseId: (callback: (leaseId: string) => Promise<void>) => Promise<void>;
    waitForStop: () => Promise<void>;
    renew: () => Promise<void>;
    track: (tenant: string, resource: string[], read: boolean) => Promise<void>;
    untrack: (tenant: string, resource: string[], read: boolean) => Promise<void>;
}

interface Lock {
    tenant: string;
    resource: string[];
}

type CreateLeaseResponse =
    | {
          error: null;
          leaseId: string;
      }
    | {
          error: ServiceError;
      };

export const createLease = async (
    connection: Connection,
    config: ClientConfig,
    serviceClient: ServiceClient
): Promise<Lease> => {
    const mutex = new Mutex();
    const ttl = 20;
    let locks: Lock[] = [];
    let rLocks: Lock[] = [];
    let leaseId = "";

    const [waitForStopExecution, stopExecution] = await createResolvablePromise<void>();
    const [waitForStop, stopped] = await createResolvablePromise<void>();

    const renew = async () => {
        const [waitForCreateLease, onCreateLease] =
            await createResolvablePromise<CreateLeaseResponse>();

        try {
            const release = await mutex.acquire();
            serviceClient.createLease(
                {
                    ip: getOwnIpAddress(),
                    app: config.appPrefix,
                    ttl,
                    leaseId,
                    alreadyRegisteredLocks: getPbLocks(leaseId, locks),
                    alreadyRegisteredRlocks: getPbLocks(leaseId, rLocks),
                },
                (error, response) => {
                    if (error) {
                        onCreateLease({ error });
                        return;
                    }

                    onCreateLease({ error: null, leaseId: response.leaseId });
                }
            );

            const response = await waitForCreateLease();

            if (response.error) {
                release();
                throw response.error;
            }

            leaseId = response.leaseId;
            keepLeaseAlive(); // do not await this!
            release();
            await connection.connect();
        } catch (e) {
            if (e === E_CANCELED) {
                return;
            }
            throw e;
        }
    };

    const keepLeaseAlive = async () => {
        while (true) {
            const response = await racePromises({
                stop: waitForStopExecution(),
                keepalive: wait((ttl / 2) * 1000),
            });

            if (response === "stop") {
                await runDropLease();
                return;
            }

            if (!(await runKeepalive())) {
                return;
            }
        }
    };

    // returns true if processing can be continued
    const runKeepalive = async () => {
        const [waitForKeepalive, onKeepalive] =
            await createResolvablePromise<ServiceError | null>();

        serviceClient.keepLease({ leaseId, ttl }, onKeepalive);

        const error = await waitForKeepalive();

        if (error) {
            await connection.disconnect();

            try {
                await retry(renew, 100, 50);
                return false;
            } catch {
                throw new Error("Unable to renew lease");
            }
        }

        return true;
    };

    const runDropLease = async () => {
        const [waitForDropped, onDrop] = await createResolvablePromise<ServiceError | null>();

        serviceClient.dropLease({ leaseId }, onDrop);

        const error = await waitForDropped();

        if (error) {
            throw error;
        }

        stopped();
    };

    return {
        stop: () => {
            mutex.cancel();
            stopExecution();
        },
        runWithLeaseId: async (callback: (leaseId: string) => Promise<void>) => {
            await mutex.runExclusive(async () => {
                await callback(leaseId);
            });
        },
        renew,
        waitForStop,
        track: async (tenant: string, resource: string[], read: boolean) => {
            try {
                await mutex.runExclusive(() => {
                    if (read) {
                        rLocks.push({ tenant, resource });
                        return;
                    }

                    locks.push({ tenant, resource });
                });
            } catch (e) {
                if (e === E_CANCELED) {
                    return;
                }
                throw e;
            }
        },
        untrack: async (tenant: string, resource: string[], read: boolean) => {
            try {
                await mutex.runExclusive(() => {
                    if (read) {
                        let found = false;
                        const newRLocks: Lock[] = [];

                        rLocks.forEach(lock => {
                            if (
                                !found &&
                                lock.tenant === tenant &&
                                arraysEqual(lock.resource, resource)
                            ) {
                                found = true;
                                return;
                            }

                            newRLocks.push(lock);
                        });

                        rLocks = newRLocks;
                        return;
                    }

                    locks = locks.filter(lock => {
                        return lock.tenant !== tenant || !arraysEqual(lock.resource, resource);
                    });
                });
            } catch (e) {
                if (e === E_CANCELED) {
                    return;
                }
                throw e;
            }
        },
    };
};

const arraysEqual = <T>(a: T[], b: T[]): boolean => {
    if (a.length !== b.length) {
        return false;
    }

    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return false;
        }
    }

    return true;
};

const getPbLocks = (leaseId: string, locks: Lock[]): PbLock[] => {
    return locks.map(lock => {
        return {
            leaseId,
            tenantId: lock.tenant,
            resource: lock.resource,
        };
    });
};
